_posts/2020-07-21-cube-functionality-to-kartothek.markdown

```diff
@@ -0,0 +1,414 @@
+Kartothek already provides Dataset features, As a user do I really need a multiple-dataset? Hang On!! 
+Let us spend rest of our time understanding the story of **Cube (kartothek that supports multiple datasets)**. 
+
+## What is a Cube?
+A Cube deals with multiple Kartothek datasets.
+
+Let us start with building a cube for **geodata**. Similar to Kartothek, 
+we need a [simplekv](https://simplekv.readthedocs.io/)-based store backend along with an abstract cube definition.
+
+**df_weather** is a pandas dataframe created from reading a csv file.
+
+```python
+>>> from io import StringIO
+>>> df_weather = pd.read_csv(
+...     filepath_or_buffer=StringIO("""
+... avg_temp     city country        day
+...        6  Hamburg      DE 2018-01-01
+...        5  Hamburg      DE 2018-01-02
+...        8  Dresden      DE 2018-01-01
+...        4  Dresden      DE 2018-01-02
+...        6   London      UK 2018-01-01
+...        8   London      UK 2018-01-02
+...     """.strip()),
+...     delim_whitespace=True,
+...     parse_dates=["day"],
+... )
+```
+
+
+```python
+>>> from kartothek.core.cube.cube import Cube
+>>>##we are creating a geodata cube instance
+>>> cube = Cube(
+...     uuid_prefix="geodata",
+...     dimension_columns=["city", "day"],
+...     partition_columns=["country"]
+...)
+```
+
+We use the simple **kartothek.io.eager_cube** backend to store the data:
+
+```python
+>>> from kartothek.io.eager_cube import build_cube
+>>> datasets_build = build_cube(
+...   data=df_weather,
+...   store=store,
+...   cube=cube
+...)
+```
+
+
+where **store** is the **simplekv** store of storefactory. (For more details, please refer our [Kartothek](2019-05-28-introducing-kartothek.markdown) blog.)
+
+We just have preserved a single Kartothek dataset. Lets print the content of seed dataset.
+
+```python
+>>> print(", ".join(sorted(datasets_build.keys())))
+seed
+>>> ds_seed = datasets_build["seed"].load_all_indices(store)
+>>> print(ds_seed.uuid)
+geodata++seed
+>>> print(", ".join(sorted(ds_seed.indices)))
+city, country, day
+```
+
+Finally, let’s have a quick look at the store content. Note that we cut out UUIDs and timestamps here for documentation purposes:
+
+```python
+>>> import re
+>>> def print_filetree(s, prefix=""):
+...     entries = []
+...     for k in sorted(s.keys(prefix)):
+...         k = re.sub("[a-z0-9]{32}", "<uuid>", k)
+...         k = re.sub("[0-9]{4}-[0-9]{2}-[0-9]{2}((%20)|(T))[0-9]{2}%3A[0-9]{2}%3A[0-9]+.[0-9]{6}", "<ts>", k)
+...         entries.append(k)
+...     print("\n".join(sorted(entries)))
+>>> print_filetree(store)
+geodata++seed.by-dataset-metadata.json
+geodata++seed/indices/city/<ts>.by-dataset-index.parquet
+geodata++seed/indices/day/<ts>.by-dataset-index.parquet
+geodata++seed/table/_common_metadata
+geodata++seed/table/country=DE/<uuid>.parquet
+geodata++seed/table/country=UK/<uuid>.parquet
+```
+
+Thus, A cube can be visualised as a metadataset of multiple datasets as shown in below image.
```

Comment:
We don't really use the term _metadataset_ anywhere in the documentation, do we? I think I would prefer something simpler like "a collection of datasets" since the _meta_ prefix is quite heavily loaded in kartothek with various different meanings

Reply:
Updated.